src/canvasObject.js

```diff
@@ -26,111 +44,88 @@ var CanvasObject = function(config, dispatcher) {
-
-            viewer.removeHandler('tile-drawn', tileDrawnHandler);
-            main.setOpacity(0, true);
-            self._fade(main, 1);
-            self._setMainImage(main);
-
-            if(previousImageObj){
-              viewer.world.removeItem(previousImageObj);
-            }
-            self.dispatcher.emit('detail-tile-source-opened', { 'detail': self.id });
-          }
-        };
-        viewer.addHandler('tile-drawn', tileDrawnHandler);
-      },
-
-      error: function(event) {
-        var errorInfo = {
-          id: self.id,
-          message: event.message,
-          source: event.source
-        };
-        self.dispatcher.emit('detail-tile-source-failed', {'detail': errorInfo});
+    var onTileDrawn = function(event) {
+      var main = event.tiledImage;
+      main.setOpacity(0, true);
+      self._fade(main, 1);
+
+      if(self.thumbnailImage){
+        viewer.world.removeItem(self.thumbnailImage);
+        self.thumbnailImage = null;
       }
-    });
+      self.fullyOpened = true;
+      self.dispatcher.emit('detail-tile-source-opened', { 'detail': self.id });
+    };
+    this.images[0].openTileSource(viewer, this.bounds, onTileDrawn);
   },
 
   openThumbnail: function(viewer) {
     var self = this;
-    viewer.addTiledImage({
-      x: this.x,
-      y: this.y,
-      width: this.width,
-      tileSource: this.placeholder,
-      opacity: this.opacity,
-      clip: this.clipRegion,
-      success: function(event) {
-        self._setMainImage(event.item);
-      }
-    })
+    var onTileDrawn = function(event) {
+      self.thumbnailImage = event.tiledImage;
+      self.dispatcher.emit('detail-thumbnail-opened', { 'detail': self.id });
+    }
+    this.thumbnail.openTileSource(viewer, this.bounds, onTileDrawn);
   },
 
   //Assumes that the point parameter is already in viewport coordinates.
   containsPoint: function(point) {
-    var rectRight = this.x + this.width;
-    var rectBottom = this.y + this.height;
+    var rectRight = this.bounds.x + this.bounds.width;
+    var rectBottom = this.bounds.y + this.bounds.height;
 
-    return (this.x <= point.x && rectRight >= point.x && this.y <= point.y && rectBottom >= point.y);
+    return (this.bounds.x <= point.x && rectRight >= point.x && this.bounds.y <= point.y && rectBottom >= point.y);
+  },
+
+  getVisibleImages: function() {
+    return this.images.filter(function(image) { return image.visible === true; });
+  },
+
+  getDetailImages: function() {
+    return this.images.filter(function(image) { return image.imageType === "detail" });
+  },
+
+  getAlternateImages: function() {
+    return this.images.filter(function(image) { return image.imageType === "alternate" });
   },
 
   setPosition: function(x, y) {
-    this.x = x;
-    this.y = y;
+    var self = this;
+    this.bounds.x = x;
+    this.bounds.y = y;
 
-    if(this.hasImageObject()) {
-      this._mainImageObj.setPosition(new OpenSeadragon.Point(x, y), true);
+    if(this.fullyOpened) {
```

Comment:
We should probably update all images regardless of whether they are shown, so they're correct later when they do need to be shown. 

This is one place where having the thumbnail be part of the images array would streamline things.